File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/board/screen/BoardViewModel.kt
@file:OptIn(ExperimentalMaterial3Api::class)

package com.ssafy.ssabree.features.board.screen

import android.util.Log
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.BoardRepository
import com.ssafy.ssabree.core.repository.PostRepository
import com.ssafy.ssabree.features.board.model.PostUiModel
import com.ssafy.ssabree.features.board.model.toUiModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

private const val TAG = "BoardViewModel"

data class BoardFilterOption(
    val id: Long?,
    val label: String
) {
    companion object {
        val All = BoardFilterOption(null, "전체보기")
    }
}

data class BoardUiState(
    val selectedFilter: BoardFilterOption = BoardFilterOption.All,
    val filterMenuExpanded: Boolean = false,
    val posts: List<PostUiModel> = emptyList(),
    val filterOptions: List<BoardFilterOption> = listOf(BoardFilterOption.All),
    val nextCursor: String? = null,
    val hasNext: Boolean = true,
    val isRefreshing: Boolean = false,
    val isLoadingMore: Boolean = false,
    val errorMessage: String? = null,
    val selectedBoardName: String = "게시판",
    val isHotSelected: Boolean = false,
    val noticeContent: String? = null
)

class BoardViewModel(
    private val postRepository: PostRepository,
    private val boardRepository: BoardRepository,
    private val initialBoardId: Long? = null // initialBoardId 추가
) : ViewModel() {

    private val pageSize = 20
    private val _uiState = MutableStateFlow(BoardUiState())
    val uiState: StateFlow<BoardUiState> = _uiState.asStateFlow()
    private var lastHotToggleTime = 0L

    init {
        refreshBoardsAndPosts(initialBoardId)
    }

    fun refreshBoardsAndPosts(boardIdToSelect: Long? = null) = viewModelScope.launch {
        _uiState.update { it.copy(isRefreshing = true, errorMessage = null) }

        val boardsResult = boardRepository.getBoards()
        val noticeResult = boardRepository.getNotice()
        val filterOptions = listOf(BoardFilterOption.All) + boardsResult.getOrElse { emptyList() }
            .map { BoardFilterOption(id = it.id, label = it.name) }

        val selected = filterOptions.find { it.id == boardIdToSelect }
            ?: filterOptions.firstOrNull()
            ?: BoardFilterOption.All
        val hotSelected = false

        _uiState.update {
            it.copy(
                filterOptions = filterOptions,
                selectedFilter = selected,
                selectedBoardName = selected.label,
                isHotSelected = hotSelected,
                posts = emptyList(),
                nextCursor = null,
                hasNext = true,
                noticeContent = noticeResult.getOrNull()
            )
        }

        loadPosts(selected.id, cursor = null, append = false)
    }

    private suspend fun loadPosts(boardId: Long?, cursor: String?, append: Boolean) {
        if (append) {
            _uiState.update { it.copy(isLoadingMore = true, errorMessage = null) }
        } else {
            _uiState.update { it.copy(isRefreshing = true, errorMessage = null) }
        }

        postRepository.getPosts(boardId = boardId, cursor = cursor, limit = pageSize)
            .onSuccess { page ->
                Log.d(TAG, "loadPosts: succeed (append=$append)")
                _uiState.update { state ->
                    val newPosts = if (append) {
                        state.posts + page.posts.map { post -> post.toUiModel() }
                    } else {
                        page.posts.map { post -> post.toUiModel() }
                    }
                    state.copy(
                        posts = newPosts,
                        nextCursor = page.nextCursor,
                        hasNext = page.hasNext,
                        isRefreshing = false,
                        isLoadingMore = false,
                        errorMessage = null
                    )
                }
            }
            .onFailure { e ->
                Log.d(TAG, "loadPosts: failed (${e.message})")
                _uiState.update {
                    it.copy(
                        isRefreshing = false,
                        isLoadingMore = false,
                        errorMessage = e.message
                    )
                }
            }
    }

    private suspend fun loadHotPosts(cursor: String?, append: Boolean) {
        if (append) {
            _uiState.update { it.copy(isLoadingMore = true, errorMessage = null) }
        } else {
            _uiState.update { it.copy(isRefreshing = true, errorMessage = null) }
        }

        postRepository.getHotPosts(cursor = cursor, limit = pageSize)
            .onSuccess { page ->
                Log.d(TAG, "loadHotPosts: succeed (append=$append)")
                _uiState.update { state ->
                    val newPosts = if (append) {
                        state.posts + page.posts.map { post -> post.toUiModel() }
                    } else {
                        page.posts.map { post -> post.toUiModel() }
                    }
                    state.copy(
                        posts = newPosts,
                        nextCursor = page.nextCursor,
                        hasNext = page.hasNext,
                        isRefreshing = false,
                        isLoadingMore = false,
                        errorMessage = null
                    )
                }
            }
            .onFailure { e ->
                Log.d(TAG, "loadHotPosts: failed (${e.message})")
                _uiState.update {
                    it.copy(
                        isRefreshing = false,
                        isLoadingMore = false,
                        errorMessage = e.message
                    )
                }
            }
    }

    fun onRefresh(boardIdToSelect: Long? = null) {
        viewModelScope.launch {
            val currentState = _uiState.value
            val isHotMode = currentState.isHotSelected

            if (isHotMode) {
                // hot 모드에서는 boardId 파라미터를 무시하고 항상 hot API만 호출
                loadHotPosts(cursor = null, append = false)
            } else {
                val targetBoardId = boardIdToSelect ?: currentState.selectedFilter.id
                _uiState.update {
                    it.copy(
                        isHotSelected = false,
                        selectedBoardName = currentState.filterOptions.find { option -> option.id == targetBoardId }?.label
                            ?: it.selectedBoardName
                    )
                }
                loadPosts(boardId = targetBoardId, cursor = null, append = false)
            }
        }
    }

    fun onFilterChipClick() {
        _uiState.update { it.copy(filterMenuExpanded = true) }
    }

    fun onFilterMenuDismiss() {
        _uiState.update { it.copy(filterMenuExpanded = false) }
    }

    fun onFilterSelected(filter: BoardFilterOption) {
        _uiState.update {
            it.copy(
                selectedFilter = filter,
                selectedBoardName = filter.label,
                isHotSelected = false,
                filterMenuExpanded = false,
                isRefreshing = true,
                posts = emptyList(),
                nextCursor = null,
                hasNext = true
            )
        }
        viewModelScope.launch {
            loadPosts(filter.id, cursor = null, append = false)
        }
    }

    fun onHotSelected() {
        val now = System.currentTimeMillis()
        if (now - lastHotToggleTime < 600) return // debounce to reduce rapid reloads
        lastHotToggleTime = now

        val state = _uiState.value

        // 이미 Hot이면 해제하고 현재 선택된 게시판으로 복귀
        if (state.isHotSelected) {
            _uiState.update {
                it.copy(
                    isHotSelected = false,
                    selectedBoardName = state.selectedFilter.label,
                    posts = emptyList(),
                    nextCursor = null,
                    hasNext = true,
                    isRefreshing = true
                )
            }
            viewModelScope.launch {
                loadPosts(state.selectedFilter.id, cursor = null, append = false)
            }
            return
        }

        _uiState.update {
            it.copy(
                isHotSelected = true,
                selectedBoardName = "Hot 게시글",
                filterMenuExpanded = false,
                posts = emptyList(),
                nextCursor = null,
                hasNext = true,
                isRefreshing = true
            )
        }
        viewModelScope.launch {
            loadHotPosts(cursor = null, append = false)
        }
    }

    fun onListEndReached(index: Int) {
        val state = _uiState.value
        val shouldLoadMore = state.hasNext &&
                !state.isLoadingMore &&
                !state.isRefreshing &&
                index >= state.posts.size - 3

        if (shouldLoadMore) {
            viewModelScope.launch {
                if (state.isHotSelected) {
                    loadHotPosts(cursor = state.nextCursor, append = true)
                } else {
                    loadPosts(state.selectedFilter.id, state.nextCursor, append = true)
                }
            }
        }
    }
}

File: iOS/ssabree/ssabree/Features/Board/BoardViewModel.swift
import Foundation
import Observation

// MARK: - Filter Option

struct BoardFilterOption: Equatable {
    let id: Int?
    let label: String

    static let all = BoardFilterOption(id: nil, label: "전체보기")
}

// MARK: - UI State

struct BoardUiState {
    var selectedFilter: BoardFilterOption = .all
    var filterMenuExpanded: Bool = false
    var posts: [PostModel] = []
    var filterOptions: [BoardFilterOption] = [.all]
    var nextCursor: String? = nil
    var hasNext: Bool = true
    var isRefreshing: Bool = false
    var isLoadingMore: Bool = false
    var errorMessage: String? = nil
    var selectedBoardName: String = "게시판"
    var isHotSelected: Bool = false
    var noticeContent: String? = nil
}

// MARK: - ViewModel

@Observable
@MainActor
final class BoardViewModel {
    private let boardRepository: BoardRepository
    private let postRepository: PostRepository
    private let pageSize = 20

    private(set) var uiState = BoardUiState()
    private var hasLoadedInitialData = false
    private let initialBoardId: Int?
    private var lastHotToggleTime: CFAbsoluteTime = 0

    init(boardRepository: BoardRepository, postRepository: PostRepository, initialBoardId: Int? = nil) {
        self.boardRepository = boardRepository
        self.postRepository = postRepository
        self.initialBoardId = initialBoardId
        // init에서 Task 제거 - View의 .task modifier에서 loadInitialDataIfNeeded 호출
    }

    /// View의 .task modifier에서 호출 - 최초 1회만 데이터 로드
    /// Task.detached를 사용하여 SwiftUI의 .task cancellation으로부터 보호
    func loadInitialDataIfNeeded() async {
        guard !hasLoadedInitialData else { return }
        hasLoadedInitialData = true

        // Task.detached: 부모 Task의 cancellation을 상속받지 않음
        // SwiftUI .task가 cancel되어도 네트워크 요청은 계속 진행됨
        await Task.detached { @MainActor [self] in
            // 게시판 목록 + 공지사항 로드
            async let boardsTask: () = self.loadBoardOptions()
            async let noticeTask: () = self.loadNotice()
            _ = await (boardsTask, noticeTask)
            // 기본적으로 전체 게시글 로드 (isHotSelected = false 이므로)
            await self.loadPosts(boardId: nil, cursor: nil, append: false)
        }.value
    }

    private func loadBoardOptions() async {
        let boardsResult = await boardRepository.getBoards()
        var filterOptions: [BoardFilterOption] = [.all]

        switch boardsResult {
        case .success(let boards):
            filterOptions += boards.map { BoardFilterOption(id: $0.id, label: $0.title) }
        case .failure(let error):
            print("[BoardViewModel] getBoards failed: \(error)")
        }

        uiState.filterOptions = filterOptions
    }

    private func loadNotice() async {
        let result = await boardRepository.getNotice()
        switch result {
        case .success(let content):
            uiState.noticeContent = content
        case .failure(let error):
            print("[BoardViewModel] getNotice failed: \(error)")
        }
    }

    func refreshBoardsAndPosts(boardIdToSelect: Int? = nil) async {
        uiState.isRefreshing = true
        uiState.errorMessage = nil

        async let boardsTask: () = loadBoardOptions()
        async let noticeTask: () = loadNotice()
        _ = await (boardsTask, noticeTask)

        let selected = uiState.filterOptions.first { $0.id == boardIdToSelect }
            ?? uiState.filterOptions.first
            ?? .all

        uiState.selectedFilter = selected
        uiState.selectedBoardName = selected.label
        uiState.isHotSelected = false
        uiState.posts = []
        uiState.nextCursor = nil
        uiState.hasNext = true

        await loadPosts(boardId: selected.id, cursor: nil, append: false)
    }

    private func loadPosts(boardId: Int?, cursor: String?, append: Bool) async {
        if append {
            uiState.isLoadingMore = true
        } else {
            uiState.isRefreshing = true
        }
        uiState.errorMessage = nil

        let result = await postRepository.getPosts(boardId: boardId, keyword: nil, cursor: cursor, limit: pageSize)

        switch result {
        case .success(let page):
            if append {
                uiState.posts += page.data
            } else {
                uiState.posts = page.data
            }
            uiState.nextCursor = page.nextCursor
            uiState.hasNext = page.hasNext
            uiState.isRefreshing = false
            uiState.isLoadingMore = false
            uiState.errorMessage = nil

        case .failure(let error):
            print("[BoardViewModel] loadPosts failed: \(error)")
            uiState.isRefreshing = false
            uiState.isLoadingMore = false
            uiState.errorMessage = error.localizedDescription
        }
    }

    private func loadHotPosts(cursor: String?, append: Bool) async {
        if append {
            uiState.isLoadingMore = true
        } else {
            uiState.isRefreshing = true
        }
        uiState.errorMessage = nil

        let result = await postRepository.getHotPosts(cursor: cursor, limit: pageSize)

        switch result {
        case .success(let page):
            if append {
                uiState.posts += page.data
            } else {
                uiState.posts = page.data
            }
            uiState.nextCursor = page.nextCursor
            uiState.hasNext = page.hasNext
            uiState.isRefreshing = false
            uiState.isLoadingMore = false
            uiState.errorMessage = nil

        case .failure(let error):
            print("[BoardViewModel] loadHotPosts failed: \(error)")
            uiState.isRefreshing = false
            uiState.isLoadingMore = false
            uiState.errorMessage = error.localizedDescription
        }
    }

    /// Pull-to-refresh: 현재 선택된 게시판의 게시글만 다시 로드
    /// Task.detached로 .refreshable의 cancellation으로부터 보호
    func onRefresh() async {
        await Task.detached { @MainActor [self] in
            self.uiState.posts = []
            self.uiState.nextCursor = nil
            self.uiState.hasNext = true

            if self.uiState.isHotSelected {
                await self.loadHotPosts(cursor: nil, append: false)
            } else {
                await self.loadPosts(boardId: self.uiState.selectedFilter.id, cursor: nil, append: false)
            }
        }.value
    }

    func onHotSelected() {
        // 600ms 디바운스
        let now = CFAbsoluteTimeGetCurrent()
        if now - lastHotToggleTime < 0.6 { return }
        lastHotToggleTime = now

        // 이미 Hot이면 해제하고 현재 선택된 게시판으로 복귀
        if uiState.isHotSelected {
            uiState.isHotSelected = false
            uiState.selectedBoardName = uiState.selectedFilter.label
            uiState.posts = []
            uiState.nextCursor = nil
            uiState.hasNext = true
            uiState.isRefreshing = true

            Task.detached { @MainActor [self] in
                await self.loadPosts(boardId: self.uiState.selectedFilter.id, cursor: nil, append: false)
            }
            return
        }

        uiState.isHotSelected = true
        uiState.selectedBoardName = "Hot 게시글"
        uiState.posts = []
        uiState.nextCursor = nil
        uiState.hasNext = true
        uiState.isRefreshing = true

        Task.detached { @MainActor [self] in
            await self.loadHotPosts(cursor: nil, append: false)
        }
    }

    func onFilterChipClick() {
        uiState.filterMenuExpanded = true
    }

    func onFilterMenuDismiss() {
        uiState.filterMenuExpanded = false
    }

    func onFilterSelected(filter: BoardFilterOption) {
        uiState.isHotSelected = false
        uiState.selectedFilter = filter
        uiState.selectedBoardName = filter.label
        uiState.filterMenuExpanded = false
        uiState.isRefreshing = true
        uiState.posts = []
        uiState.nextCursor = nil
        uiState.hasNext = true

        Task.detached { @MainActor [self] in
            await self.loadPosts(boardId: filter.id, cursor: nil, append: false)
        }
    }

    func onListEndReached(index: Int) {
        let shouldLoadMore = uiState.hasNext &&
            !uiState.isLoadingMore &&
            !uiState.isRefreshing &&
            index >= uiState.posts.count - 3

        if shouldLoadMore {
            Task.detached { @MainActor [self] in
                if self.uiState.isHotSelected {
                    await self.loadHotPosts(cursor: self.uiState.nextCursor, append: true)
                } else {
                    await self.loadPosts(boardId: self.uiState.selectedFilter.id, cursor: self.uiState.nextCursor, append: true)
                }
            }
        }
    }
}
